specification/sbi_atee.adoc

```diff
@@ -13,6 +44,953 @@ running a VMM).
+both measured and code pages that are part of the TVM payload, and confidential zero-pages that can be
+mapped in by the VMM on demand. The non-confidential regions include non-confidential memory that can
+be used for virtual device emulation and so forth.
+
+=== Operational model
+
+Executing confidential workloads in a TEE requires a sequence of one or more of the steps detailed below.
+We'll assume that these steps are performed by an untrusted entity like the OS/VMM in conjunction with the
+TSM.
+
+. Platform TSM detection and capability enumeration
+. Conversion of non-confidential memory to confidential memory
+. Trusted VM (TVM) creation
+. Donating confidential memory to the TSM for TVM page management
+. Defining confidential memory regions
+. Defining MMIO regions
+. Defining non-confidential shared memory regions
+. Conversion of non-confidential (V)CPUs to confidential-VCPUs
+. Mapping code and data payloads for the TVM to confidential-memory regions
+. Mapping demand-zero confidential memory regions
+. Adding confidential VCPUs to the TVM
+. Finalizing TVM creation
+. Scheduling TVM execution
+. Management of secure interrupts
+. Handling and servicing TVM faults and exits
+. Mapping of confidential zero pages on demand
+. Mapping non-confidential shared pages on demand
+. Tearing down TVMs
+. Repurposing confidential memory for other TVMs
+. Reclaiming confidential memory for non-confidential VMs
+
+=== Platform TSM detection and capability enumeration
+
+Platform support for the TSM can be detected by probing for the EXT_TEE extension, and then
+calling sbi_tee_tsm_get_info() to get information about the current status of the TSM. The
+TSM must be be in TSM_READY in order to process further ECALLs.
+
+=== Conversion of non-confidential memory to confidential memory
+
+Platform memory is non-confidential by default, and must be converted to confidential memory
+before use with TVMs. The conversion process is initiated by designating the host physical
+pages that to be converted, and then issuing fence operations to ensure that all outstanding
+TLB entries to the non-confidential memory are flushed across all VCPUs on the platform. This
+ensures that there's no overlapping mapping between the confidential and non-confidential memory
+regions on the platform.
+
+This requires the host to make three separate ECALLs to the TSM:
+
+. sbi_tee_tsm_convert_pages()
+. sbi_tee_tsm_initiate_fence()
+. sbi_tee_tsm_local_fence()
+
+The memory conversion process is complete when sbi_tee_tsm_local_fence() is successfully completed
+on the every VCPU on the platform.
+
+Converted memory may be assigned to TVMs, but cannot be repurposed for regular use unless it's reclaimed.
+
+=== TVM creation
+
+TVMs are created using the sbi_tee_create_tvm(). This creates a TVM, and sets with state "initializing". The
+host must assign confidential memory for page tables, payload mapping, and VCPUs before it can be transitioned
+into a "runnable" state.
+
```

Comment:
Table for all states of TVM. It would be good to add a state transition diagram too. But that can be added later. Don't want to burden you with so much documentation ;)

Reply:
+1 (also should the TVM states be enumerated as all caps like the TSM states - in this case, TVM_INITIALIZING, TVM_RUNNABLE?

Reply:
Should we also add an API to enumerate the TVM state?

Reply:
probably a NTH

Reply:
Sorry, not familiar with the NTH TLA :).

Reply:
nice to have